Feature/button control

GrzegorzWojnarowicz opened the pull request:
https://github.com/darts-plus/darts-plus-embedded/issues/2#issuecomment-1072752458

Keep that in mind

GrzegorzWojnarowicz committed:
#2 Hardware button handler

add Button class

GrzegorzWojnarowicz committed:
#2 Hardware button handler

alternative version of button_pressed method

GrzegorzWojnarowicz committed:
#2 Hardware button handler

fix button_pressed method

Selvan66 commented on lib/button-control/src/button.cpp:
```diff
@@ -0,0 +1,58 @@
+      unsigned long time_diff = release_time - press_time < 0 ? 0 : release_time - press_time; 
+      output = time_diff > long_press ? PRESSED_LONG :
+              ((time_diff > medium_press) && (time_diff <= long_press)) ? PRESSED_MEDIUM :
+              ((time_diff > 0) && (time_diff <= medium_press)) ? PRESSED_SHORT : NOT_PRESSED;
```
You can split this to `if` but i don't see too much problem with this if its stay in that form. 
But what i would change for sure you can cut this `if` because if  `time_diff > long_press` is false it's for sure `time_diff <= long_press` so you don't need to write it in else statement. Same to `medium_press`.

delipl replied:
Yes, it looks really complicated and too long.

delipl replied:
You can create a separate private methods for some `if` statements, for example: `check_long()`, `check_short()`, `check_medium()` etc.

xHeler commented on lib/button-control/inc/button.hpp:
```diff
@@ -0,0 +1,28 @@
+const unsigned long medium_press = 1000;
+const unsigned long long_press = 5000;
+
+enum Press {
```
Press what ? Maybe change to -> PressState or smth give more information

delipl commented on lib/button-control/inc/button.hpp:
```diff
@@ -0,0 +1,28 @@
+
+#include <Arduino.h>
+const unsigned long debounce_delay = 75;
+const int high_button_state = HIGH;
```
unnecessary use just `HIGH`

xHeler commented on lib/button-control/inc/button.hpp:
```diff
@@ -0,0 +1,28 @@
+class Button
+{
+private:
+    short pinNum;
+    unsigned long press_time;
+    unsigned long release_time;
+    bool button_active;
+    unsigned long time_diff;
```
Two different name convention. Ones you use camelCase in the other hand you use snake_case

delipl commented on lib/button-control/inc/button.hpp:
```diff
@@ -0,0 +1,28 @@
+    PRESSED_LONG = 3
+};
+
+class Button
```
Please move the `public` things just under class name.
It is easier to read and use then the public methods are just under the class name.

xHeler commented on lib/button-control/src/button.cpp:
```diff
@@ -0,0 +1,58 @@
+
+}
+
+Button::~Button()
+{
+}
```
Button::~Button() {} - in one line looks more sexy

delipl commented on lib/button-control/inc/button.hpp:
```diff
@@ -0,0 +1,28 @@
+const unsigned long medium_press = 1000;
+const unsigned long long_press = 5000;
+
+enum Press {
```
please add `unknown` state :D

you can also use `enum class`

delipl replied:
If the values of `enum` are in sequence you don't have to assign the value.

bartoszkwiat reviewed (commented):
Clean code

delipl commented on lib/button-control/inc/button.hpp:
```diff
@@ -0,0 +1,28 @@
+class Button
+{
+private:
+    short pinNum;
```
- There are two types of the cases. `pinNum` and `press_time` . Chose one and implement to others :D `camelCase` or `snake_case`

- Use explicite types of `ints`. https://en.cppreference.com/w/cpp/types/integer . The `pinNum` can be `uint8_t` - unsigned int 8 bit. 0 - 255.

delipl commented on lib/button-control/inc/button.hpp:
```diff
@@ -0,0 +1,28 @@
+    unsigned long time_diff;
+public:
+    Button(int buttonNum);
+    ~Button();
```
If the destructor is empty just don't use it ;)

GrzegorzWojnarowicz committed:
#2 Hardware button handler

change pressed_button method from millis to interrupts

change from class Button to namespace ns_button

Selvan66 reviewed (commented):
I liked the solution in the class more than in the namespace, but if others think it might stay that way, I have no problem with that.

Just write some example program to use this code.

GrzegorzWojnarowicz commented:
It is in namespace now just because interrupts had problems with classes. The problem is that when interrupt happens it calls a function (in my class it was a method, and it wouldn't know which instance of this method to call). I've seen that someone told that it may be done by static methods and address them using switch case, but he said that it's not a perfect solution.

xHeler commented on lib/button-control/inc/button.hpp:
```diff
@@ -0,0 +1,99 @@
+    const unsigned long medium_press = 1000;
+    const unsigned long long_press = 5000;
+    
+    enum Button_state {
```
Why not put them all and create button class ?

delipl replied:
Because we need static functions as callbacks to the interrupts.
I think that the buttons have to be something static.

Selvan66 replied:
I looked at it and it turns out that you can use lambda expressions in `attachInterrupt()`, you only need to attach the library ` #include <FunctionalInterrupt.h> ` from esp

xHeler commented:
I'd like to have examples directory with simple example of usage your buttons. 
And please create README.md with information about library.

delipl reviewed (changes requested):
Add this pattern to function with argument. Please split this to `.hpp` and `.cpp`
``` cpp
if (press_time[i] != 0)
        {
            if ((millis() - press_time[i]) > debounce_delay)
            {
                release_time[i] = millis();
            }
        }else{
            if ((millis() - release_time[i]) > debounce_delay)
            {
                press_time[i] = millis();
            }
        }
    ```

GrzegorzWojnarowicz committed:
#2 Hardware button handler

change button_pressed function into set_current_state

add get_current_state function

use as state-machine using get_current_state function

GrzegorzWojnarowicz committed:
#2 Hardware button handler

add README file with class description and usage example
